File: src/vibesnake/core/near_miss.py
"""Detect and classify close spatial and timing outcomes.

The detector emits warning or scored events for body proximity, constrained
routes, boundary travel, late food collection, and boosted collection. Its
thresholds and bonuses are provisional balance parameters. A cooldown prevents
stationary or oscillating play from producing an unbounded event stream.
"""

from typing import Tuple, Set
from dataclasses import dataclass
from vibesnake.data import settings


@dataclass(frozen=True)
class NearMissEvent:
    """Describe one warning or scored proximity event.

    A position of ``(-1, -1)`` marks an event that is not tied to one grid
    cell. Warning events do not award points or start the detector cooldown.
    """

    type: str  # Event category identifier
    position: Tuple[int, int]  # Grid location (or -1,-1 for non-spatial)
    score_bonus: int  # Points added when the event is rewarded
    message: str  # Player feedback text
    color: Tuple[int, int, int]  # RGB for visual notification
    is_warning: bool = False  # If True, visual-only warning (no bonus/cooldown)


class NearMissDetector:
    """Emit bounded near-miss events from current run geometry and timers."""

    def __init__(self):
        self.recent_events = []  # Track recent near-misses for combo
        self.event_timeout = 3.0  # Seconds before events expire
        self.last_near_miss_time = 0.0  # Cooldown to prevent spam
        self.near_miss_cooldown = 1.5  # Minimum seconds between near-miss events

    def update(self, dt: float):
        """Update and expire old events."""
        self.recent_events = [(event, time - dt) for event, time in self.recent_events if time - dt > 0]
        # Update cooldown timer
        if self.last_near_miss_time > 0:
            self.last_near_miss_time -= dt

    def check_near_miss(
        self, head: Tuple[int, int], body_positions: Set[Tuple[int, int]], snake_length: int
    ) -> NearMissEvent | None:
        """Classify body occupancy in the eight cells around the head.

        Snakes shorter than eight segments are ignored. Two occupied neighbors
        emit an unscored warning. Three emit a one-point event and four or more
        emit a two-point event. Scored events share a 1.5-second cooldown.
        These thresholds remain provisional balance parameters.
        """
        # Skip if snake is too short (no danger possible)
        if snake_length < 8:  # Require longer snake for meaningful danger
            return None

        # Check all 8 surrounding cells
        hx, hy = head
        adjacent_cells = [
            (hx + dx, hy + dy)
            for dx in [-1, 0, 1]
            for dy in [-1, 0, 1]
            if not (dx == 0 and dy == 0)  # Skip the head itself
        ]

        # Count how many adjacent cells have body segments
        danger_count = sum(1 for cell in adjacent_cells if cell in body_positions)

        # Four or more occupied neighbors produce the higher-value event.
        if danger_count >= 4:
            # Only trigger if cooldown expired (prevent spam)
            if self.last_near_miss_time > 0:
                return None

            bonus = 2
            self.last_near_miss_time = self.near_miss_cooldown  # Start cooldown
            return NearMissEvent(
                type="near_miss",
                position=head,
                score_bonus=bonus,
                message="THREADING THE NEEDLE!",
                color=(255, 100, 255),  # Bright purple
                is_warning=False,  # Full event with bonus
            )

        # Tier 2: Near-miss - 3 sides dangerous (CLOSE CALL)
        elif danger_count >= 3:
            # Only trigger if cooldown expired (prevent spam)
            if self.last_near_miss_time > 0:
                return None

            bonus = 1  # Standard near-miss bonus
            self.last_near_miss_time = self.near_miss_cooldown  # Start cooldown
            return NearMissEvent(
                type="near_miss",
                position=head,
                score_bonus=bonus,
                message="CLOSE CALL!",
                color=(255, 200, 0),  # Gold
                is_warning=False,  # Full event with bonus
            )

        # Tier 1: Pre-warning - 2 sides dangerous (VISUAL ONLY)
        elif danger_count == 2:
            # No cooldown check - pre-warnings can display continuously
            return NearMissEvent(
                type="danger_warning",
                position=head,
                score_bonus=0,  # No bonus for pre-warning
                message="",  # No message (visual feedback only)
                color=(255, 50, 50),  # Red glow for danger
                is_warning=True,  # Flag as warning (no cooldown trigger)
            )

        return None

    def check_edge_ride(
        self, head: Tuple[int, int], direction: Tuple[int, int], snake_length: int
    ) -> NearMissEvent | None:
        """Reward motion parallel to a wrapping boundary.

        The provisional bonus is ``snake_length // 10`` clamped to 1 through
        10. No retained dataset currently establishes that length is an
        accurate proxy for the difficulty of this action.
        """
        hx, hy = head
        dx, dy = direction

        # Check if at edge AND moving parallel to it
        at_left = hx == 0 and dy != 0
        at_right = hx == settings.GRID_WIDTH - 1 and dy != 0
        at_top = hy == 0 and dx != 0
        at_bottom = hy == settings.GRID_HEIGHT - 1 and dx != 0

        if at_left or at_right or at_top or at_bottom:
            # Scale the provisional reward with length and cap score inflation.
            bonus = min(max(snake_length // 10, 1), 10)

            # Select the message from the same deterministic bonus thresholds.
            if bonus >= 8:
                message = "EDGE MASTERY!"
            elif bonus >= 5:
                message = "EDGE LORD!"
            else:
                message = "EDGE RIDE"

            return NearMissEvent(
                type="edge_ride",
                position=head,
                score_bonus=bonus,
                message=message,
                color=(100, 255, 255),  # Cyan
            )

        return None

    def check_clutch_eat(self, starvation_timer: float, starvation_max: float) -> NearMissEvent | None:
        """
        Check if food was eaten with very little time left.

        The 1.5-second window is provisional. Seeded telemetry must measure its
        trigger distribution, and human review must determine whether the cue is
        legible and earned before the threshold is treated as release tuning.

        Args:
            starvation_timer: Current starvation timer (counts up from 0)
            starvation_max: Maximum time before death (typically 30.0s)

        Returns:
            NearMissEvent if clutch moment detected (time_remaining < 1.5s)
        """
        time_remaining = starvation_max - starvation_timer

        # Tightened threshold: 1.5s (was 3.0s)
        if time_remaining < 1.5:
            bonus = 1  # Just +1 for clutch moments
            return NearMissEvent(
                type="clutch_eat",
                position=(-1, -1),  # No specific position
                score_bonus=bonus,
                message="CLUTCH!",
                color=(255, 50, 50),  # Red
            )

        return None

    def check_style_points(self, has_boost: bool) -> NearMissEvent | None:
        """
        Check if food eaten while boosting (style points).

        Args:
            has_boost: Whether boost power-up is active

        Returns:
            NearMissEvent if style moment detected
        """
        if has_boost:
            return NearMissEvent(
                type="style_points",
                position=(-1, -1),
                score_bonus=1,  # Just +1 for style
                message="ZOOMING!",
                color=(255, 150, 0),  # Orange
            )

        return None

    def add_event(self, event: NearMissEvent):
        """Track event for combo purposes."""
        self.recent_events.append((event, self.event_timeout))

    def get_combo_multiplier(self) -> float:
        """Get bonus multiplier based on recent near-miss combo."""
        event_count = len(self.recent_events)
        if event_count >= 3:
            return 2.0  # Triple near-miss = 2x multiplier
        elif event_count >= 2:
            return 1.5
        return 1.0

File: native/src/VibeSnake.Rules/NearMiss.cs
namespace VibeSnake.Rules;

/// <summary>
/// Classification of a bounded risk-reward proximity or style event.
/// Thresholds and bonuses are provisional balance parameters.
/// </summary>
public enum NearMissKind : byte
{
    DangerWarning = 0,
    BodyProximity = 1,
    EdgeRide = 2,
    ClutchEat = 3,
    StylePoints = 4,
}

/// <summary>
/// One immutable near-miss classification. Warnings carry no score and do not
/// start the reward cooldown. A null position marks a non-spatial event.
/// </summary>
public readonly record struct NearMissEvent(
    NearMissKind Kind,
    GridPoint? Position,
    int ScoreBonus,
    string Message,
    bool IsWarning);

/// <summary>
/// Pure fixed-step near-miss detector. Cooldown and recent-event windows use
/// rules ticks so outcomes stay deterministic and independent of wall clocks.
/// </summary>
public sealed class NearMissDetector
{
    public const int MinimumSnakeLength = 8;
    public const int DefaultCooldownTicks = 30;
    public const int DefaultEventTimeoutTicks = 60;
    public const int ClutchRemainingTicks = 30;
    public const int MaximumEdgeRideBonus = 10;

    private readonly List<(NearMissEvent Event, int RemainingTicks)> _recentEvents = [];
    private int _rewardCooldownTicksRemaining;

    public NearMissDetector(
        int cooldownTicks = DefaultCooldownTicks,
        int eventTimeoutTicks = DefaultEventTimeoutTicks)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(cooldownTicks);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(eventTimeoutTicks);

        CooldownTicks = cooldownTicks;
        EventTimeoutTicks = eventTimeoutTicks;
    }

    public int CooldownTicks { get; }

    public int EventTimeoutTicks { get; }

    public int RewardCooldownTicksRemaining => _rewardCooldownTicksRemaining;

    public IReadOnlyList<NearMissEvent> RecentEvents
    {
        get
        {
            if (_recentEvents.Count == 0)
            {
                return Array.Empty<NearMissEvent>();
            }

            var copy = new NearMissEvent[_recentEvents.Count];
            for (var index = 0; index < _recentEvents.Count; index++)
            {
                copy[index] = _recentEvents[index].Event;
            }

            return copy;
        }
    }

    /// <summary>
    /// Advances cooldown and expires tracked combo events by one or more rules ticks.
    /// </summary>
    public void AdvanceTicks(int ticks = 1)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(ticks);

        if (ticks == 0)
        {
            return;
        }

        if (_rewardCooldownTicksRemaining > 0)
        {
            _rewardCooldownTicksRemaining = Math.Max(0, _rewardCooldownTicksRemaining - ticks);
        }

        if (_recentEvents.Count == 0)
        {
            return;
        }

        for (var index = _recentEvents.Count - 1; index >= 0; index--)
        {
            var remaining = _recentEvents[index].RemainingTicks - ticks;
            if (remaining <= 0)
            {
                _recentEvents.RemoveAt(index);
            }
            else
            {
                _recentEvents[index] = (_recentEvents[index].Event, remaining);
            }
        }
    }

    /// <summary>
    /// Classifies body occupancy in the eight non-wrapping cells around the head.
    /// Snakes shorter than <see cref="MinimumSnakeLength"/> are ignored.
    /// </summary>
    public NearMissEvent? CheckBodyProximity(
        GridPoint head,
        IReadOnlySet<GridPoint> bodyPositions,
        int snakeLength)
    {
        ArgumentNullException.ThrowIfNull(bodyPositions);
        if (snakeLength < MinimumSnakeLength)
        {
            return null;
        }

        var dangerCount = 0;
        for (var dx = -1; dx <= 1; dx++)
        {
            for (var dy = -1; dy <= 1; dy++)
            {
                if (dx == 0 && dy == 0)
                {
                    continue;
                }

                if (bodyPositions.Contains(new GridPoint(head.X + dx, head.Y + dy)))
                {
                    dangerCount++;
                }
            }
        }

        if (dangerCount >= 4)
        {
            if (_rewardCooldownTicksRemaining > 0)
            {
                return null;
            }

            _rewardCooldownTicksRemaining = CooldownTicks;
            return new NearMissEvent(
                NearMissKind.BodyProximity,
                head,
                ScoreBonus: 2,
                Message: "THREADING THE NEEDLE!",
                IsWarning: false);
        }

        if (dangerCount >= 3)
        {
            if (_rewardCooldownTicksRemaining > 0)
            {
                return null;
            }

            _rewardCooldownTicksRemaining = CooldownTicks;
            return new NearMissEvent(
                NearMissKind.BodyProximity,
                head,
                ScoreBonus: 1,
                Message: "CLOSE CALL!",
                IsWarning: false);
        }

        if (dangerCount == 2)
        {
            return new NearMissEvent(
                NearMissKind.DangerWarning,
                head,
                ScoreBonus: 0,
                Message: string.Empty,
                IsWarning: true);
        }

        return null;
    }

    /// <summary>
    /// Rewards motion parallel to a wrapping boundary. Bonus is
    /// <c>clamp(snakeLength / 10, 1, 10)</c>.
    /// </summary>
    [System.Diagnostics.CodeAnalysis.SuppressMessage(
        "Performance",
        "CA1822:Mark members as static",
        Justification = "The established instance member is retained for public API compatibility.")]
    public NearMissEvent? CheckEdgeRide(
        GridPoint head,
        Direction direction,
        int snakeLength,
        int gridWidth,
        int gridHeight)
    {
        if (gridWidth < 2 || gridHeight < 2)
        {
            throw new ArgumentOutOfRangeException(nameof(gridWidth), "Grid dimensions must be at least 2.");
        }

        if (snakeLength < 1)
        {
            return null;
        }

        var offset = direction.Offset();
        var atLeft = head.X == 0 && offset.Y != 0;
        var atRight = head.X == gridWidth - 1 && offset.Y != 0;
        var atTop = head.Y == 0 && offset.X != 0;
        var atBottom = head.Y == gridHeight - 1 && offset.X != 0;
        if (!(atLeft || atRight || atTop || atBottom))
        {
            return null;
        }

        var bonus = Math.Clamp(snakeLength / 10, 1, MaximumEdgeRideBonus);
        var message = bonus switch
        {
            >= 8 => "EDGE MASTERY!",
            >= 5 => "EDGE LORD!",
            _ => "EDGE RIDE",
        };

        return new NearMissEvent(
            NearMissKind.EdgeRide,
            head,
            ScoreBonus: bonus,
            Message: message,
            IsWarning: false);
    }

    /// <summary>
    /// Rewards food collection when fewer than <see cref="ClutchRemainingTicks"/>
    /// hunger ticks remain.
    /// </summary>
    [System.Diagnostics.CodeAnalysis.SuppressMessage(
        "Performance",
        "CA1822:Mark members as static",
        Justification = "The established instance member is retained for public API compatibility.")]
    public NearMissEvent? CheckClutchEat(int hungerTicksRemaining)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(hungerTicksRemaining);

        if (hungerTicksRemaining >= ClutchRemainingTicks)
        {
            return null;
        }

        return new NearMissEvent(
            NearMissKind.ClutchEat,
            Position: null,
            ScoreBonus: 1,
            Message: "CLUTCH!",
            IsWarning: false);
    }

    /// <summary>
    /// Rewards food collection while Boost is active.
    /// </summary>
    [System.Diagnostics.CodeAnalysis.SuppressMessage(
        "Performance",
        "CA1822:Mark members as static",
        Justification = "The established instance member is retained for public API compatibility.")]
    public NearMissEvent? CheckStylePoints(bool hasBoost)
    {
        if (!hasBoost)
        {
            return null;
        }

        return new NearMissEvent(
            NearMissKind.StylePoints,
            Position: null,
            ScoreBonus: 1,
            Message: "ZOOMING!",
            IsWarning: false);
    }

    public void TrackEvent(NearMissEvent nearMissEvent)
    {
        if (nearMissEvent.IsWarning)
        {
            return;
        }

        _recentEvents.Add((nearMissEvent, EventTimeoutTicks));
    }

    /// <summary>
    /// Bounded multiplier from recent rewarded near-miss events.
    /// Two events yield 1.5x; three or more yield 2.0x.
    /// </summary>
    public double GetComboMultiplier()
    {
        var count = _recentEvents.Count;
        if (count >= 3)
        {
            return 2.0;
        }

        if (count >= 2)
        {
            return 1.5;
        }

        return 1.0;
    }

    public void Reset()
    {
        _recentEvents.Clear();
        _rewardCooldownTicksRemaining = 0;
    }
}
